output/types.go

```diff
@@ -75,10 +75,17 @@ type WithTestRunStop interface {
 	SetTestRunStopCallback(func(error))
 }
 
-// WithRunStatusUpdates means the output can receive test run status updates.
-type WithRunStatusUpdates interface {
+// WithStopWithTestError allows output to receive the error value that the test
+// finished with. It could be nil, if the test finished nominally.
+//
+// If this interface is implemented by the output, StopWithError() will be
+// called instead of Stop().
+//
+// TODO: refactor the main interface to use this method instead of Stop()? Or
+// something else along the lines of https://github.com/grafana/k6/issues/2430 ?
+type WithStopWithTestError interface {
```

Comment:
I think the proposal in #2430 that outputs should stop when the `samples` channel is closed would be cleaner, but we would still need a way for them to query the test error, as suggested [here](https://github.com/grafana/k6/issues/2430#issuecomment-1318540637).

So WDYT if we add a method to `output.Params` for this purpose? This would avoid this deprecation of `Stop()` and eventual breaking change, and only outputs that need it can call it. Eventually, we can deprecate and remove `Stop()` in favor of the proposal in #2430, but that would be a single breaking change, instead of the two in this case.

An alternative to changing `output.Params` could be for outputs to subscribe to error events, using the event system suggested for JS extensions in #2432, but I think that's still a long ways off.

Reply:
We don't have a breaking change here, this is why I added an extra optional interface. All of the existing outputs and output extensions that use the old `Stop()` method will continue to work. They will just continue not knowing if the test finished with an error or not, which most outputs don't care about anyway.

I am softly against adding a control surface to `output.Params`, it is counter-intuitive. But I am open to discussion and willing to be convinced otherwise, when we decide to tackle https://github.com/grafana/k6/issues/2430. For now, this solution allows us to delay any breaking changes until we do so, exactly so we can do them only once and not constantly break extensions.

Reply:
We're avoiding a breaking change here by doing this optional expansion of the `Output` interface, which is IMO a code smell. If we decide to remove `Stop()` in #2430, then we'll also need to remove `StopWithTestError()`, hence two eventual breaking changes. I'd rather we avoided adding future work, and reconsider how outputs can query test errors. Essentially, doing this in a way that brings us closer to #2430, not farther.

Adding a method to `output.Params` was a suggestion, but I also agree not a very intuitive one. Maybe via some other public method?

Reply:
> which is IMO a code smell

Why? I'd consider it a fairly idiomatic Go thing, the stdlib does it all the time :confused: 

> If we decide to remove Stop() in https://github.com/grafana/k6/issues/2430, then we'll also need to remove StopWithTestError(), hence two eventual breaking changes.

It is likely that when we tackle https://github.com/grafana/k6/issues/2430, we will refactor [all of the existing interfaces](https://github.com/grafana/k6/blob/41472d1ba41737be699fc756ad0cdec4b2f0540f/output/types.go#L18-L88), and that won't be 6 different breaking changes, it will be one big change :sweat_smile: And even so, it's likely it will be somewhat graceful, since we will likely have a Go shim to the old `Output` APIs that will work (with a warning) for a few k6 versions.

In any case, we still don't have a clear idea of how exactly we will do https://github.com/grafana/k6/issues/2430, and when exactly we will do it, or even *if* we will actually do it... So I don't want to block this PR and its follow-ups until we have that figured out. And, FWIW, figuring out how the new Outputs will look like should come after we have removed the `Engine` in https://github.com/grafana/k6/pull/2815, since it changes things in this area a lot, but it also depends on this PR first.

Reply:
OK, let's leave this for later then.

> Why? I'd consider it a fairly idiomatic Go thing, the stdlib does it all the time

I don't mean in the general sense. Composing interfaces is sometimes clean and convenient. But in this case it means that outputs don't just implement `output.Output`, but some outputs also implement `WithStopWithTestError`, which besides being awkwardly named, also requires runtime type checks wherever this additional functionality is required. And we're doing all this just to avoid a breaking change, which could be prevented by rethinking how this is done, and getting us closer to #2430.